lib/pdk/context.rb

```diff
@@ -0,0 +1,97 @@
+require 'pdk'
+
+module PDK
+  module Context
+    autoload :None, 'pdk/context/none'
+    autoload :Module, 'pdk/context/module'
+    autoload :ControlRepo, 'pdk/context/control_repo'
+
+    # Automatically determines the PDK Context given a path. Create will continue up the directory tree until it
+    # finds a valid context
+    # @return [PDK::Context::AbstractContext] Returns a PDK::Context::None if the context could not be determined
+    def self.create(context_path)
+      return PDK::Context::None.new(context_path) unless PDK::Util::Filesystem.directory?(context_path)
+
+      previous = nil
+      current = PDK::Util::Filesystem.expand_path(context_path)
+      until !PDK::Util::Filesystem.directory?(current) || current == previous
+        # Control Repo detection
+        return PDK::Context::ControlRepo.new(current, context_path) if PDK::ControlRepo.control_repo_root?(current)
+
+        # Puppet Module detection
+        # Note - The metadata.json file check should really be in PDK::Util.in_module_root?
```

Comment:
This was done intentionally so that modules without a `metadata.json` could be converted.

Reply:
Ok.  Will remove the comment

Reply:
Hrmm... so `.in_module_root?` only checks for common directories. So adding a check for metadata.json wouldn't affect that scenario.

In fact in a empty directory with only `metadata.json`:
`PDK::Util.module_root` will return the module path but
`PDK::Util.in_module_root?(..)` for that same path will return false which is "odd"

Reply:
Removed the comment.  Can debate this elsewhere.